functest/main_test.go

```diff
@@ -0,0 +1,266 @@
+/*
+ *    Copyright 2018 Insolar
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package functest
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"go/build"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/insolar/insolar/logicrunner/goplugin/testutil"
+	"github.com/stretchr/testify/assert"
+)
+
+const HOST = "http://localhost:19191"
+const TestUrl = HOST + "/api/v1"
+const insolarImportPath = "github.com/insolar/insolar"
+
+func testdataPath() string {
+	p, err := build.Default.Import("github.com/insolar/insolar", "", build.FindOnly)
+	if err != nil {
+		panic(err)
+	}
+	return filepath.Join(p.Dir, "testdata", "functional")
+}
+
+func buildInsolard() (string, error) {
+	insolardPath := filepath.Join(testdataPath(), "insolard")
+	_, err := exec.Command(
+		"go", "build",
+		"-o", insolardPath,
+		insolarImportPath+"/cmd/insolard/",
+	).CombinedOutput()
+	if err != nil {
+		return "", err
+	}
+	return insolardPath, nil
+}
+
+// Create dir for runnercodepath, build ginsider-cli for logic runner correct start, and finally build insolard
+// Return path to binary and err
+func setup(currentFile string) (string, error) {
+	err := os.MkdirAll(filepath.Join(filepath.Dir(currentFile), "contractstorage"), 0777)
+	if err != nil {
+		return "", err
+	}
+
+	_, _, err = testutil.Build()
+	if err != nil {
+		return "", err
+	}
+
+	return buildInsolard()
+}
+
+func wait(stdout io.ReadCloser) {
+	scanner := bufio.NewScanner(stdout)
+	for scanner.Scan() {
+		line := scanner.Text()
+		fmt.Println(line)
+		if strings.Contains(line, "======= Host info ======") {
+			break
+		}
+	}
+}
+
+func exitFromInsolard(stdin io.WriteCloser) {
+	io.WriteString(stdin, "exit\n")
+}
+
+func teardown(cmd *exec.Cmd, currentFile string, stdin io.WriteCloser) {
+	exitFromInsolard(stdin)
+	err := cmd.Wait()
+	if err != nil {
+		fmt.Println("try to kill, wait done with error: ", err)
+		err := cmd.Process.Kill()
+		if err != nil {
+			fmt.Println("failed to kill process: ", err)
+		}
+	}
+	err = os.RemoveAll(filepath.Join(filepath.Dir(currentFile), "data"))
```

Comment:
filepath.Join(filepath.Dir(currentFile) - let's move to var, since it's used more than one time